### PR TITLE
Solidity contract to mint aBTC draft PR

### LINKED ISSUE #35: Solidity contract to mint aBTC
# Description

Solidity contract to mint aBTC with a one-to-one mint with wBTC, tBTC, cbBTC and more.

# TODO 
- [] Architecture and Doc of the implementation
- [] Mapping of BTC wrapped like tBTC, wBTC, cbBTC etc.
- [] ERC20 or ERC4626 aBTC with Access control from OpenZepellin: https://docs.openzeppelin.com/contracts/5.x/erc4626
- [] Deposit function
-[] Withdraw function
- [] Cross call into Cairo
- [] Upgradeable contract if possible

### PR DESCRIPTION
**Pull request by PoulavBhowmick03**
Fixes #35

### CODE REVIEW

File: onchain/solidity_contracts/src/defi/ABTCVault.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        userDeposits[msg.sender][_wrappedBTCToken] += _amount;
+        
+        emit Deposited(msg.sender, _wrappedBTCToken, _amount, aBTCAmount);
+    }
+
+    function withdraw(address _wrappedBTCToken, uint256 _aBTCAmount) external whenNotPaused {
+        WrappedBTC memory wrappedToken = wrappedBTCTokens[_wrappedBTCToken];
+        require(wrappedToken.isPermitted, "Token not permitted");
+        
+        uint256 wrappedBTCAmount = (_aBTCAmount * 1e18) / wrappedToken.ratio;
```
Comment (MSghais) [inline onchain/solidity_contracts/src/defi/ABTCVault.sol:76]: Can we change it without ratio atm please?
We are gonna add a fees percentage after as a ratio for a protocol fee.
Can be setup to 0 to do the test, or just comment it

File: onchain/solidity_contracts/src/defi/ABTCVault.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        require(_ratio > 0, "Invalid ratio");
+        wrappedBTCTokens[_token] = WrappedBTC(_isPermitted, _ratio, _poolingTimestamp);
+        emit TokenPermissionSet(_token, _isPermitted, _ratio, _poolingTimestamp);
+    }
+
+    function deposit(address _wrappedBTCToken, uint256 _amount) external whenNotPaused {
+        WrappedBTC memory wrappedToken = wrappedBTCTokens[_wrappedBTCToken];
+        require(wrappedToken.isPermitted, "Token not permitted");
+        
+        uint256 aBTCAmount = (_amount * wrappedToken.ratio) / 1e18;
```
Comment (MSghais) [inline onchain/solidity_contracts/src/defi/ABTCVault.sol:62]: Can we change it without ratio atm please?
We are gonna add a fees percentage after as a ratio for a protocol fee.
Can be setup to 0 to do the test, or just comment it

File: onchain/solidity_contracts/test/ABTCVaultTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        assertTrue(vault.hasRole(vault.MINTER_ROLE(), admin));
+        assertTrue(vault.hasRole(vault.UPGRADER_ROLE(), admin));
+        assertTrue(vault.hasRole(vault.PAUSER_ROLE(), admin));
+    }
+
+    function testSetWrappedBTCToken() public {
+        vault.setWrappedBTCToken(address(wbtc), true, 1e18, block.timestamp);
+        (bool isPermitted, uint256 ratio, uint256 poolingTimestamp) = vault.wrappedBTCTokens(address(wbtc));
+        assertTrue(isPermitted);
+        assertEq(ratio, 1e18);
```
Comment (MSghais) [inline onchain/solidity_contracts/test/ABTCVaultTest.t.sol:54]: Can we change the ratio as a fee percentage?
Setup to 0 first and we can change it after

**Review by MSghais — CHANGES_REQUESTED** [review]

**Comment by PoulavBhowmick03** [comment]
@MSghais Made the changes. PTAL

**Pull request closed without merging**